### DISCUSSION TITLE
Mojo and Causal Based AI

### DISCUSSION THREAD

**Discussion by taxles** [modular/modular#591] [Ideas]
I couldn't see any mention of this in Mojo's [Roadmap](https://docs.modular.com/mojo/roadmap.html ) nor [Documentation](https://docs.modular.com/) so (as requested in the roadmap) I've raised  a new discussion here for my question.

Q. Is there any intention in Mojo's future to develop constructs to declare Cause & Effect Relationships between two variables (where such relationships could be a (new) Type), and then put these relationships into a Structural Equation Model (SEM) based on a Directed Acyclic Graph (DAG)?

These are a few foundational constructs needed for implementing Judea Pearl's Causal Based AI.

**Comment by lattner** [comment]
This is outside of my area of experience, but it sounds like something that would be best built as a library and not into the language itself.  We'd love for you to explore this, but we're not tracking any specific implementation work, so I'll close this out
